File: week_2/day_08_stack/next_greater_follow_up_questions/stockSpan.js
function stockSpan(prices) {
    const stack = [];
    const span = new Array(prices.length);

    for (let i = 0; i < prices.length; i++) {
        while (stack.length && prices[stack[stack.length - 1]] <= prices[i]) {
            stack.pop();
        }
        span[i] = stack.length === 0 ? i + 1 : i - stack[stack.length - 1];
        stack.push(i);
    }
    return span;
}

// INPUT
console.log(stockSpan([100,80,60,70,60,75,85])); // [1,1,1,2,1,4,6]

File: week_2/day_08_stack/next_greater_follow_up_questions/stock_span.py
def stockSpan(prices):
    stack = []
    span = [0] * len(prices)

    for i in range(len(prices)):
        while stack and prices[stack[-1]] <= prices[i]:
            stack.pop()
        span[i] = i + 1 if not stack else i - stack[-1]
        stack.append(i)

    return span

# INPUT
prices = [100, 80, 60, 70, 60, 75, 85]
print(stockSpan(prices))  # Output: [1,1,1,2,1,4,6]
